### PR TITLE
add unit test for chapter attribution

### PR DESCRIPTION
**Pull request by zednis**
Not complete.  Still need to add the qualified attribution via the REST API.

references USGCRP/gcis-ontology/issues/113

### CODE REVIEW

**Comment by zednis** [comment]
@bduggan How do I add the qualified attribution to the the unit test model?

**Comment by bduggan** [comment]
Send a POST request to /:resource/contributors.  Here's an example:
  https://github.com/USGCRP/gcis/blob/master/t/018_contributor.t#L82

**Comment by zednis** [comment]
@bduggan The script failed at https://github.com/zednis/gcis/blob/patch-12/t/011_sparql.t#L206 with a 404.  Do you know why?

**Comment by bduggan** [comment]
The primary key for person is "id", not "identifier", so just change `id` to `identifier` in the POST to /person.
Also to see errors like this (and status codes other than 200 for failures), send a json accept header.  This can be done for all of these requests by adding this at the top of the test:
```
$t->ua->on(start => sub { pop->req->headers->header("Accept" => "application/json")});
```
Also to diagnose errors, this is handy:
```
diag $t->tx->res->body;
````
which will produce output in verbose mode, e.g.
```
./Build test --test-files=t/011_sparql.t --verbose=1
```
(when doing this I saw an error when posting to .../contributor since the organization with identifier "test" does not exist)

File: t/011_sparql.t [hunk 1]
```diff
@@ -190,6 +213,10 @@ add_to_model("/image/$image_identifier");
 add_to_model("/dataset/$dataset_identifier");
 add_to_model("/platform/$platform_identifier");
 add_to_model("/instrument/$instrument_identifier");
+add_to_model("/role_type/$role_identifier");
+add_to_model("/person/$person_identifier");
+add_to_model("/organization/$org_identifier");
+add_to_model("/report/trees/chapter/contributors/the-larch");
```
Comment (zednis) [inline t/011_sparql.t]: @bduggan do you have any idea why this is failing?

Comment (bduggan) [inline t/011_sparql.t]: The URI ```/report/<report>/chapter/contributors/<chapter>```  does not provide turtle in response to a GET request with a .ttl suffix or an Accept heade.  This should instead just be a URI like ```/report/trees/chapter/the-larch``` which will provide turtle (which is generated using the corresponding turtle template).

**Comment by justgo129** [comment]
@zednis please let me know when you're able to act on the suggestion of @bduggan and test accordingly.  I can then merge the pull request.

**Comment by zednis** [comment]
@bduggan I appear to have fixed the syntactic errors but my sparql query is not returning the expected results so the test is still failing.

Is there anything wrong with this POST to add the contributor relationship?
```
$t->post_ok("/report/trees/chapter/contributors/the-larch" => json => { person_id => $person_identifier, organization_identifier => $org_identifier, role => $role_identifier })->status_is(200);
```

and if not, what am I missing on add_to_model to ensure this RDF is added to the in-memory model that is queried?

**Comment by rewolfe** [comment]
It looks okay.  See:

https://github.com/USGCRP/gcis/blob/master/t/017_person_api.t

Line 64:

$t->post_ok("/report/contributors/uno" => json =>
    {person_id => $id, organization_identifier => 'earth', role =>
'engineer' })->status_is(200);

Could you give the actual values you are using?  Note: person_id is numeric.

On Tue, Oct 13, 2015 at 1:20 PM, Stephan Zednik <notifications@github.com>
wrote:

> @bduggan <https://github.com/bduggan> I appear to have fixed the
> syntactic errors but my sparql query is not returning the expected results
> so the test is still failing.
>
> Is there anything wrong with this POST to add the contributor relationship?
>
> $t->post_ok("/report/trees/chapter/contributors/the-larch" => json => { person_id => $person_identifier, organization_identifier => $org_identifier, role => $role_identifier })->status_is(200);
>
> and if not, what am I missing on add_to_model to ensure this RDF is added
> to the in-memory model that is queried?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGCRP/gcis/pull/244#issuecomment-147782845>.
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by zednis** [comment]
```
my $person_identifier = "101";
my $org_identifier = "acme";
my $role_identifier = "lead_author";
```

**Comment by rewolfe** [comment]
I'm just guessing here, but try putting "contributors" right after
"report", e.g.

  /report/contributors/trees/chapter/the-larch

On Tue, Oct 13, 2015 at 4:49 PM, Stephan Zednik <notifications@github.com>
wrote:

> my $person_identifier = "101";
> my $org_identifier = "acme";
> my $role_identifier = "lead_author";
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGCRP/gcis/pull/244#issuecomment-147848906>.
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by zednis** [comment]
@bduggan is @rewolfe correct?  should chapter contributors be `` /report/contributors/trees/chapter/the-larch`` or ``/report/trees/chapter/contributors/the-larch`` ?

**Comment by rewolfe** [comment]
Stephan, I believe this works.  Take a look at:

   https://github.com/USGCRP/gcis-scripts/blob/master/lib/Gcis/Exim.pm#L1361

On Tue, Oct 13, 2015 at 6:07 PM, Stephan Zednik <notifications@github.com>
wrote:

> @bduggan <https://github.com/bduggan> is @rewolfe
> <https://github.com/rewolfe> correct? should chapter contributors be
> /report/contributors/trees/chapter/the-larch or
> /report/trees/chapter/contributors/the-larch ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGCRP/gcis/pull/244#issuecomment-147868084>.
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by justgo129** [comment]
@zednis please let me know when you've resolved it.  At that time, I'll merge this pull request as well as https://github.com/USGCRP/gcis-ontology/issues/113.

File: t/011_sparql.t [hunk 2]
```diff
@@ -179,6 +179,27 @@ my $dbpedia_platform_identifier = "Three-pod";
...
+my $org_identifier = "acme";
+$t->post_ok(
+  "/organization" => json => {
+    identifier  => $org_identifier,
+    name        => "Acme Inc.",
+  }
+)->status_is(200);
+
+# The chapter was attributed to that author
+$t->post_ok("/report/trees/chapter/contributors/the-larch" => json =>
```
Comment (rewolfe) [inline t/011_sparql.t]: @zednis - Try changing this to: /report/contributors/trees/chapter/the-larch

Comment (zednis) [inline t/011_sparql.t]: thanks @rewolfe

Comment (rewolfe) [inline t/011_sparql.t]: @zednis - Not quite there.   The other problem is that
organization_identifier has to be the full uri: /organization/acme

On Tue, Dec 1, 2015 at 11:47 AM, Stephan Zednik <notifications@github.com>
wrote:

> In t/011_sparql.t
> <https://github.com/USGCRP/gcis/pull/244#discussion_r46303352>:
>
> > +
> > +# add a person
> > +my $person_identifier = "101";
> > +$t->post_ok( "/person" => json => { id => $person_identifier, first_name  => "John", last_name   => "Doe", } )->status_is(200);
> > +
> > +# add an organization
> > +my $org_identifier = "acme";
> > +$t->post_ok(
> > +  "/organization" => json => {
> > +    identifier  => $org_identifier,
> > +    name        => "Acme Inc.",
> > +  }
> > +)->status_is(200);
> > +
> > +# The chapter was attributed to that author
> > +$t->post_ok("/report/trees/chapter/contributors/the-larch" => json =>
>
> thanks @rewolfe <https://github.com/rewolfe>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGCRP/gcis/pull/244/files#r46303352>.
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

Comment (rewolfe) [inline t/011_sparql.t]: @zednis - Oh, and I think the original location of contributors was okay.
The url should be

/report/trees/chapter/contributors/the-larch

Sorry for the confusion.

On Tue, Dec 1, 2015 at 11:58 AM, Robert Wolfe <rewolfe@usgcrp.gov> wrote:

> @zednis - Not quite there.   The other problem is that
> organization_identifier has to be the full uri: /organization/acme
>
> On Tue, Dec 1, 2015 at 11:47 AM, Stephan Zednik <notifications@github.com>
> wrote:
>
>> In t/011_sparql.t
>> <https://github.com/USGCRP/gcis/pull/244#discussion_r46303352>:
>>
>> > +
>> > +# add a person
>> > +my $person_identifier = "101";
>> > +$t->post_ok( "/person" => json => { id => $person_identifier, first_name  => "John", last_name   => "Doe", } )->status_is(200);
>> > +
>> > +# add an organization
>> > +my $org_identifier = "acme";
>> > +$t->post_ok(
>> > +  "/organization" => json => {
>> > +    identifier  => $org_identifier,
>> > +    name        => "Acme Inc.",
>> > +  }
>> > +)->status_is(200);
>> > +
>> > +# The chapter was attributed to that author
>> > +$t->post_ok("/report/trees/chapter/contributors/the-larch" => json =>
>>
>> thanks @rewolfe <https://github.com/rewolfe>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/USGCRP/gcis/pull/244/files#r46303352>.
>>
>
>
>
> --
> Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

Comment (zednis) [inline t/011_sparql.t]: I think I am confused on your comment re: org identifier, are you saying the organization_identifier attribute on line 201 should in effect be this?

organization_identifier => "/organization/acme"

This does not appear to be consistent with the person_id attribute.  Or is that wrong as well?

Comment (rewolfe) [inline t/011_sparql.t]: @zednis - Correct, the organization_identifier should be /organization/acme
on line 201.  The person is the just the number, not the complete uri,
whereas the organziation is the uri.

On Tue, Dec 1, 2015 at 12:22 PM, Stephan Zednik <notifications@github.com>
wrote:

> In t/011_sparql.t
> <https://github.com/USGCRP/gcis/pull/244#discussion_r46308368>:
>
> > +
> > +# add a person
> > +my $person_identifier = "101";
> > +$t->post_ok( "/person" => json => { id => $person_identifier, first_name  => "John", last_name   => "Doe", } )->status_is(200);
> > +
> > +# add an organization
> > +my $org_identifier = "acme";
> > +$t->post_ok(
> > +  "/organization" => json => {
> > +    identifier  => $org_identifier,
> > +    name        => "Acme Inc.",
> > +  }
> > +)->status_is(200);
> > +
> > +# The chapter was attributed to that author
> > +$t->post_ok("/report/trees/chapter/contributors/the-larch" => json =>
>
> I think I am confused on your comment re: org identifier, are you saying
> the organization_identifier attribute on line 201 should in effect be this?
>
> organization_identifier => "/organization/acme"
>
> This does not appear to be consistent with the person_id attribute. Or is
> that wrong as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/USGCRP/gcis/pull/244/files#r46308368>.
>



-- 
Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by rewolfe** [comment]
Weird, now it is failing at line 70?

#   Failed test 'got one row'
#   at t/011_sparql.t line 70.

**Comment by zednis** [comment]
perhaps because I had commented out the add_to_model line?

``add_to_model("/report/trees/chapter/contributors/the-larch");``

**Comment by rewolfe** [comment]
@zednis - try changing line L383 to from "rdf:type <http://xmlns.com/foaf/0.1/Person>" to just "gcis:person"
see: http://data.globalchange.gov/person/840.thtml

**Comment by justgo129** [comment]
cd /usr/local/projects/release/tuba
   ./bin/tuba_stop_httpd  # stops the database
    git pull github master
    git push repos
    ./bin/tuba_start_httpd
 
    # If the turtle templates have been updated, rebuild virtuoso like so:
    ssh gcis-dev-rdf
    cd rdf-scripts
    ./import-gcis

**Comment by justgo129** [comment]
@zednis did that work?

**Comment by zednis** [comment]
@justgo129 no, it did not.  I am ready to close this ticket as not done.

**Comment by lomky** [comment]
@zednis I'm catching up on backlog gcis pull requests; any change since January or should I close as not done?

**Comment by zednis** [comment]
I think we gave up on getting this test to work.  You can close as not done.

**Pull request closed without merging**